backends/dpdk/dpdkAsmOpt.cpp

```diff
@@ -264,5 +264,236 @@ bool ValidateTableKeys::preorder(const IR::DpdkAsmProgram *p) {
+    usesInfo[b->src1->toString()]++;
+    usesInfo[b->src2->toString()]++;
+    defInfo[b->dst->toString()]++;
+}
+
+void CopyPropagationAndElimination::markUseDef(const IR::DpdkMovStatement *mv) {
+    defInfo[mv->dst->toString()]++;
+    usesInfo[mv->src->toString()]++;
+    replacementMap[mv->dst->toString()] = mv->src;
+}
+
+void CopyPropagationAndElimination::markUseDef(const IR::DpdkUnaryStatement *u) {
+    usesInfo[u->src->toString()]++;
+    defInfo[u->dst->toString()]++;
+}
+
+void CopyPropagationAndElimination::markUseDef(const IR::DpdkCastStatement *c) {
+    usesInfo[c->src->toString()]++;
+    defInfo[c->dst->toString()]++;
+    replacementMap[c->dst->toString()] = c->src;
+}
+
+void CopyPropagationAndElimination::markUseDef(const IR::DpdkJmpCondStatement *b) {
+    usesInfo[b->src1->toString()]++;
+    usesInfo[b->src2->toString()]++;
+}
+
+void CopyPropagationAndElimination::markUseDef(const IR::DpdkLearnStatement *b) {
+    usesInfo[b->timeout->toString()]++;
+    if (b->argument)
+        usesInfo[b->argument->toString()]++;
+}
+
+const IR::Expression* CopyPropagationAndElimination::replaceIfCopy(const IR::Expression *expr,
+                                                                    bool allowConst) {
+    if (!expr)
+        return expr;
+    auto str = expr->toString();
+    if (haveSingleUseDef(str)) {
+        if (auto rexpr = getIrreplaceableExpr(str, allowConst)) {
+            return rexpr;
+        }
+    }
+     if (!expr->is<IR::Constant>())
+        newUsesInfo[expr->toString()]++;
+    return expr;
+}
+
+void CopyPropagationAndElimination::elimCastOrMov(const IR::DpdkAsmStatement* stmt,
+                                        IR::IndexedVector<IR::DpdkAsmStatement>& instr) {
+    const IR::Expression* srcExpr = nullptr, *dstExpr = nullptr;
+    if (stmt->is<IR::DpdkMovStatement>()) {
+        auto mv = stmt->to<IR::DpdkMovStatement>();
+        srcExpr = mv->src;
+        dstExpr = mv->dst;
+    } else {
+        auto mv = stmt->to<IR::DpdkCastStatement>();
+        srcExpr = mv->src;
+        dstExpr = mv->dst;
+    }
+    auto src = srcExpr->toString();
+    auto dst = dstExpr->toString();
+    bool dropCopy = false;
+    if (dontEliminate.count(dst) == 0
+            && dst.startsWith("m.")
+            && newUsesInfo[dst] == 0
+            && haveSingleUseDef(dst)) {
+        dropCopy = true;  // do nothing, drop the statement
+    }
+    if (!dropCopy) {
+        auto r = replaceIfCopy(srcExpr);
+        if (dst != r->toString() && dst.startsWith("m.")) {
+            newUsesInfo[r->toString()]++;
+            instr.push_back(new IR::DpdkMovStatement(dstExpr, r));
+        } else {
+            instr.push_back(stmt);
+            newUsesInfo[src]++;
+        }
+    }
+}
+
+void CopyPropagationAndElimination::collectUseDef(IR::IndexedVector<IR::DpdkAsmStatement> stmts) {
+    for (auto stmt : stmts) {
+        if (auto mv = stmt->to<IR::DpdkMovStatement>()) {
```

Comment:
usually a visitor does this kind of dispatch.

Reply:
yes but we need ordered statement list, which dpdk backend already provides.

Reply:
visitors will traverse the list in the order of the statements inside.

Reply:
is it possible to get statements in reverse order also?

Reply:
I meant can we make visitor visit statement in reverse order ? Instead of collecting all statements in a vector and reversing it.

Reply:
no